lib/src/main/scala/code/chymyst/jc/Molecules.scala

```diff
@@ -288,51 +295,82 @@ class M[T](val name: String) extends (T => Unit) with NonblockingMolecule[T] {
+  * Reply is successful if the emitting call does not time out. In this case, we have a reply value.
+  * This is represented by [[HaveReply]].
+  * If the reply times out, there is no reply value. This is represented by [[ReplyTimedOut]].
+  * If the reaction finished but did not reply, it is an error condition. If the reaction finished and replied more than once, it is also an error condition.
+  * After a reaction fails to reply
+  */
+private[jc] sealed trait ReplyStatus {
+  def isWaiting: Boolean = false
+  def haveReply: Boolean = false
+  def isTimedOut: Boolean = false
+  def isRepeated: Boolean = false
+}
+
+private[jc] case object WaitingForReply extends ReplyStatus {
+  override def isWaiting: Boolean = true
+}
+
+/** Indicates that the reaction body finished running but did not reply to the blocking molecule.
   *
-  * result Reply value as {{{Option[R]}}}. Initially this is None, and it may be assigned at most once by the
-  *               "reply" action if the reply is "valid" (i.e. not timed out).
-  * semaphore Mutable semaphore reference. This is initialized only once when creating an instance of this
-  *                  class. The semaphore will be acquired when emitting the molecule and released by the "reply"
-  *                  action. The semaphore will be destroyed and never initialized again once a reply is received.
-  * errorMessage Optional error message, to notify the caller or to raise an exception when the user made a
-  *                     mistake in chemistry.
-  * replyTimeout Will be set to "true" if the molecule was emitted with a timeout and the timeout was reached.
-  * replyRepeated Will be set to "true" if the molecule received a reply more than once.
+  * @param message Error message (showing which other molecules did not receive replies, or received multiple replies).
+  */
+private[jc] final case class ErrorNoReply(message: String) extends ReplyStatus
+
+private[jc] case object ReplyTimedOut extends ReplyStatus {
+  override def isTimedOut: Boolean = true
+}
+
+private[jc] case object ReplyWasRepeated extends ReplyStatus {
+  override def isRepeated: Boolean = true
+}
+/** Indicates that the reaction body correctly replied to the blocking molecule.
+  *
+  * @param result The reply value.
+  * @tparam R Type of the reply value.
+  */
+private[jc] final case class HaveReply[R](result: R) extends ReplyStatus {
+  override def haveReply: Boolean = true
+}
 
+/** This trait contains the implementations of most methods for the [[ReplyValue]] and [[EmptyReplyValue]] classes.
+  *
+  *
+  * semaphore
+  * semaphoreForReplyStatus
   *
   * @tparam T Type of the value that the molecule carries.
   * @tparam R Type of the reply value.
   */
 private[jc] trait AbsReplyValue[T, R] {
 
-  @volatile var result: Option[R] = None
-
-  @volatile private var semaphore: Semaphore = {
-    val s = new Semaphore(0, false); s.drainPermits(); s
-  }
+  @volatile var replyStatus: ReplyStatus = WaitingForReply
```

Comment:
@volatile  makes it possible to modify a variable from one thread and immediately see the new value from another thread, which would otherwise require an explicit `synchronize` or another such barrier. Without @volatile, the new value of the variable could have been delayed to a future time when the local memory of the thread is written to the global memory. (The "memory barrier" of Java multithreading model.)

I could have used an `AtomicReference` here, but the logic still requires semaphores, so I decided to go with a `volatile`.